### ISSUE TITLE
Change in IMAPStore.close() may cause JVM out of memory

### ISSUE THREAD

**Issue by gbulfon** [eclipse-ee4j/mail#562]
We have recently moved to jakarta mail 2.0 from java mail 1.5.4.

We started to face sporadic problems in our mail webapp where tomcat would go 100% cpu and then go out of memory.
After many dumps and threads analysys we found that every time this was happening we would find blocked threads during finalize, like this:

Thread 2790: (state = BLOCKED)
 - java.lang.Object.wait(long) @bci=0 (Compiled frame; information may be imprecise)
 - java.lang.Object.wait() @bci=2, line=502 (Compiled frame)
 - com.sun.mail.imap.IMAPFolder.waitIfIdle() @bci=157, line=3326 (Compiled frame)
 - com.sun.mail.imap.IMAPFolder.close(boolean, boolean) @bci=71, line=1501 (Interpreted frame)
 - com.sun.mail.imap.IMAPFolder.forceClose() @bci=3, line=1469 (Interpreted frame)
 - com.sun.mail.imap.IMAPStore.closeAllFolders(boolean) @bci=114, line=1706 (Compiled frame)
 - com.sun.mail.imap.IMAPStore.close() @bci=6, line=1606 (Compiled frame)
 - com.sun.mail.imap.IMAPStore.finalize() @bci=40, line=1621 (Compiled frame)
 - java.lang.System$2.invokeFinalize(java.lang.Object) @bci=1, line=1273 (Compiled frame)
 - java.lang.ref.Finalizer.runFinalizer(sun.misc.JavaLangAccess) @bci=46, line=102 (Compiled frame)
 - java.lang.ref.Finalizer.access$100(java.lang.ref.Finalizer, sun.misc.JavaLangAccess) @bci=2, line=34 (Compiled frame)
 - java.lang.ref.Finalizer$FinalizerThread.run() @bci=45, line=217 (Compiled frame)

We then decided to check if anything was changed from java mail 1.5.4 during IMAPStore.close() and we actually found a change done in 1.6.x by Bill Shannon himself, like this:

closeAllFolders(false); => closeAllFolders(true);
emptyConnectionPool(false); => emptyConnectionPool(true);

We then remembered that using 1.5.4, a couple of years ago, we made some similar changes to force closeAllFolders(true) during close, by overriding close() in our SonicleIMAPStore instance, thinking to have better performances during session close, but discovered soon that this was causing a deadlock somewhere randomly, so we took that customization off of our code.

Now, this change in jakarta mail is putting back our change directly into the base libarary, and this is causing a serious problem.
Before deciding to fork and use a modified version of jakarta mail, I'm asking if you're willing to have a way to customize this behaviour of finalize -> close with closeAllFolders(true/false), maybe via some setter or property we may customize on our SonicleIMAPStore extension, or any other solution you may think of.

Thanks,
Gabriele

**Comment by jmehrens** [comment]
Your code might be leaking folders and stores in your code calling into the mail API.  Looking at [IMAPFolder.java#L1496](https://github.com/eclipse-ee4j/mail/blob/832654f11d7adde372f1c66ea25f838a71dc1779/mail/src/main/java/com/sun/mail/imap/IMAPFolder.java#L1496) the close is a no-op if the client code promptly closed the folder before it goes out of scope.  You could also try setting 'mail.imap.finalizecleanclose=false' and or 'mail.imaps.finalizecleanclose=false' but that should be the default anyway.

**Comment by gbulfon** [comment]
I see your point.
Actually we have a single IMAPStore for each web session, then we have our own folders cache in wrapper objects that maintain many more infos about the folder, useful for the web app (e.g. it's international name and many more).
These wrapper objects are kept inside an hashmap during the whole web session.
Folders are kept open as much as possible during the session, so that a user working only on the inbox will have this folder kept open during all the operations.
Once the user works around the folders, a pool of maximum 5 is kept open, closing the less recently used.
Once the session is closed (by the user explicitly or because of a zombie web session being destroyed by tomcat), the cleanup code takes care of destroying all our wrapper folders cache objects and hashmaps, then calls IMAPStore.close(), which we believe would take care of cleaning up javamail stuff, so closing all the opened folders: we don't call an explicit close of the folders during cleanup of our cache.

As the code inside IMAPStore.close() is calling IMAPStore.closeAllFolders(boolean), and this is done during our session close, why would an IMAPStore object to be garbage collected find a folder still open?

**Comment by jmehrens** [comment]
I think the core problem is the finalizer progress is being blocked by the fact that waitIFIdle is making the finalizer thread wait for signal that never arrives or doesn't arrive soon enough.

> As the code inside IMAPStore.close() is calling IMAPStore.closeAllFolders(boolean), and this is done during our session close, why would an IMAPStore object to be garbage collected find a folder still open?

The code may call IMAPStore.close but do you have log output that proves it was invoked when you expected it to be invoked?  I don't have an answer because we need more information.  You need to turn on debugging for IMAP (see IMAP docs).  You need to capture the logger finest output from waitIfIdle method.  That log information will help determine why the notify is not arriving.

Explicitly closing the folders first before the store is closed would be ideal and what is recommend.   Since you are running under Tomcat you have to deal with classloader maze and explicit closing of resources helps ensure that right context classloader is invoking your method.  You could also look at setting the thread context classloader to the classloader of the store object before closing and then finally resetting it after close. The finailzer thread will most likely be a parent classloaser which can't reach into the child loader.  If classloading on finalization occurs that is going to trigger a linkage error.  Might be a source of the folder being considered open.

Next steps are (one change at a time):

1. Add logging to capture that store close was invoked and returned normally.
2. Add logging to capture the context classloader used to close the store and the classloader of the store.
2. Enable logging so you can see the folder waitIfIdle statements.
3. Ensure CCE used during close is the CL of the store class.
4. Ensure all folders are closed by your code before closing the store. Set CCE to classloader of folder when doing so.
5. Experimental hackery: Set the interrupt status of the current thread before calling folder.close.
6. Experimental hackery: Set the interrupt status of the current thread before calling store.close.
